### PR TITLE
files ls

### PR DESCRIPTION
**Pull request by pgte**
Based on https://github.com/ipfs/js-ipfs/pull/940 and https://github.com/ipfs/js-ipfs-unixfs-engine/pull/197

- [x] core `ipfs.files.ls` and `ipfs.ls`
- [x] cli `files ls` and `ls` alias
- [x] http server api
- [x] http client api
- [x] interface-ipfs-core tests

### CODE REVIEW

**Comment by diasdavid** [comment]
@pgte make sure to rebase master onto this branch, there were some fixes there that will affect how your tests run (not your codebase though :))

**Comment by pgte** [comment]
Depends on

- [x] https://github.com/ipfs/interface-ipfs-core/pull/167 and
- [x] current master on ipfs-unixfs-engine

File: src/core/components/files.js [hunk 1]
```diff
@@ -13,6 +13,7 @@ const waterfall = require('async/waterfall')
 const isStream = require('is-stream')
 const Duplex = require('stream').Duplex
 const CID = require('cids')
+const toB58String = require('multihashes').toB58String
```
Comment (diasdavid) [inline src/core/components/files.js]: use `new CID(hash).toBaseEncodedString()` instead

File: src/http/api/routes/files.js [hunk 2]
```diff
@@ -42,4 +42,16 @@ module.exports = (server) => {
       validate: resources.files.add.validate
     }
   })
+
+  api.route({
+    // TODO fix method
+    method: '*',
+    path: '/api/v0/files/ls',
```
Comment (diasdavid) [inline src/http/api/routes/files.js]: Should be just `/api/v0/ls`

**Comment by diasdavid** [comment]
Tests failed for me with:

image

**Comment by diasdavid** [comment]
CI failed because of missing linting

https://travis-ci.org/ipfs/js-ipfs/jobs/301842793#L2971-L2973

I'm fixing it

**Comment by richardschneider** [comment]
I just got fresh code from master and it appears that `files ls` is still not there.  Am I doing something wrong?

```
> git checkout master
> git pull
> npm install
> node src\cli\bin.js files --help
bin.js files

Operations over files (add, cat, get, ls, etc)

Commands:
  bin.js files add <file>       Add a file to IPFS using the UnixFS data format
  bin.js files cat <ipfs-path>  Fetch and cat an IPFS path referencing a file
  bin.js files get <ipfs-path>  Fetch a file or directory with files references
                                from an IPFS Path

Options:
  --version  Show version number                                       [boolean]
  --silent   Write no output                          [boolean] [default: false]
  --help     Show help                                                 [boolean]
```

**Comment by diasdavid** [comment]
@richardschneider This implemented `ipfs.ls`. Right now in the state of things in go-ipfs there is:

- ipfs ls
- ipfs file ls
- ipfs files ls (mfs exclusive)

We want to converge all of them, step by step :)

**Pull request merged**